packages/core-sdk/resolvers/token-resolvers/resolvers/aave/src/query/functions/getTokenComponents.ts

```diff
@@ -0,0 +1,82 @@
+import { V1_LENDING_PROTOCOL_ID, V1_UNISWAP_PROTOCOL_ID } from "../constants";
+import {
+  env,
+  Ethereum_Connection,
+  Ethereum_Query,
+  Input_getTokenComponents,
+  Interface_Token,
+  Interface_TokenComponent,
+  QueryEnv,
+  Token_Query,
+  Token_Token,
+  Token_TokenType,
+} from "../w3";
+
+function fetchUnderlyingTokenAddress(
+  token: Token_Token,
+  connection: Ethereum_Connection,
+  protocolId: string,
+): string {
+  // v1 protocols use a different function name, but abi is otherwise like v2
+  const fun: string =
+    protocolId == V1_LENDING_PROTOCOL_ID || protocolId == V1_UNISWAP_PROTOCOL_ID
+      ? "underlyingAssetAddress"
+      : "UNDERLYING_ASSET_ADDRESS";
+  const res = Ethereum_Query.callContractView({
+    address: token.address,
+    method: `function ${fun}() view returns (address)`,
+    args: null,
+    connection: connection,
+  });
+  if (res.isErr) {
+    throw new Error("Invalid Aave protocol token: " + token.address);
+  }
+  return res.unwrap();
+}
+
+export function getTokenComponents(input: Input_getTokenComponents): Interface_TokenComponent {
+  if (env == null) throw new Error("env is not set");
```

Comment:
I feel like this check should be backed into our `codegen` with a function named: `ensureEnv()`